File: functional/JavaScript/caesar_cipher.js
function shifter(char, shift) {
  let ciphered_char = " ";
  // skip if whitespace
  if(char.trim() != '') {
    // Handle if Uppercase
    if (char === char.toUpperCase()) {
      ciphered_char = String.fromCharCode(((char.charCodeAt(0) - 65 + shift) % 26) + 65);
    }
    // Handle if Lowercase
    else {
      ciphered_char = String.fromCharCode(((char.charCodeAt(0) + shift - 97) % 26) + 97);
    }
  }
  return ciphered_char;
}


function encrypt(toCipher, shift) {
    const toCipher_arr = toCipher.split('');
    let shiftAmt = shift % 26;
    // functionally map each character to shifter method
    const ciphered_arr = toCipher_arr.map(char => shifter(char, shiftAmt));
    var ciphered = ciphered_arr.join("");
    return ciphered;
}


function decrypt(ciphered, shift) {
    var negativeShift = (26 - shift)
    return encrypt(ciphered, negativeShift)
}


function solve(cipher, shift) {
  console.log("Caesar " + shift + ": " + encrypt(cipher, shift));
  if (shift > 0) {
    solve(cipher, shift-1)
  }
}


/** Main **/
// Initialization
var cipherStr = "Dude Wheres My Car";
var shift = 4;

console.log("Welcome to Caesar Ciphers with JavaScript!");

// Call Caesar Cipher Encryption Method
var ciphered = encrypt(cipherStr, shift);
console.log("Encrypted Cipher Value: " + ciphered);

// Call Caesar Cipher Decryption Method
var deciphered = decrypt(ciphered, shift);
console.log("Decrypted Cipher Value: " + deciphered);

// Call Caesar Cipher Solve Method
solve("DUDE", 26);

File: functional/Scala/caesar_cipher.scala
object CaesarCipher {

    var cipherStr = ""
    
    def shifter(char: Char, shift: Int) = {
        // Account for uppercase letters
        if (char >= 'A' && char <= 'Z') {
            // If no whitespace, shift character
            if(!(char.isWhitespace)) {
                print((((char.toInt + shift - 65) % 26).toInt + 65).toChar.toString())
                this.cipherStr += (((char.toInt + shift - 65) % 26).toInt + 65).toChar.toString()
            }
            // Else whitespace, append space 
            else {
                print(" ")
                this.cipherStr += (" ")
            }
        }
        // Account for lowercase letters
        else {
            if(!(char.isWhitespace)) {
                print((((char.toInt + shift - 97) % 26).toInt + 97).toChar.toString())
                this.cipherStr += (((char.toInt + shift - 65) % 26).toInt + 65).toChar.toString()
            }
            else {
                print(" ")
                this.cipherStr += (" ")
            }
        }
    }


    def encrypt(toCipher: String, shift: Int) = {
        var shiftAmt: Int = shift % 26
        toCipher.map(char => shifter(char, shiftAmt))
	}


    def decrypt(shift: Int) = {
		encrypt(this.cipherStr, (-shift))
	}


    def solve(cipher: String, shift: Int): Unit = {
        println("Caesar " + shift + ": " + encrypt(cipher, shift))
        if(shift > 0) {
            solve(cipher, shift-1)
        }
    }

    
    def main(args: Array[String]): Unit = {
        
        // Initialization
        var cipherStr = "Dude Wheres My Car"
        var shift = 4
        
        println("Welcome to Caesar Ciphers with Scala!")
        
        // Call Caesar Cipher Encryption Method
        print("Encrypted Cipher Value: ")
        encrypt(cipherStr, shift)
        println("")
  
        // Call Caesar Cipher Decryption Method
        print("Decrypted Cipher Value: ")
        decrypt(shift)
        println("")
      
        // Call Caesar Cipher Solve Method
        solve("DUDE", 26)
    }
}
